sway/config/output.c

```diff
@@ -584,3 +480,150 @@ void free_output_config(struct output_config *oc) {
+		sway_log_errno(SWAY_ERROR, "fcntl failed");
+		return false;
+	}
+	return true;
+}
+
+static bool _spawn_swaybg(char **command) {
+	if (config->swaybg_client != NULL) {
+		wl_client_destroy(config->swaybg_client);
+	}
+	int sockets[2];
+	if (socketpair(AF_UNIX, SOCK_STREAM, 0, sockets) != 0) {
+		sway_log_errno(SWAY_ERROR, "socketpair failed");
+		return false;
+	}
+	if (!set_cloexec(sockets[0], true) || !set_cloexec(sockets[1], true)) {
+		return false;
+	}
+
+	config->swaybg_client = wl_client_create(server.wl_display, sockets[0]);
+	if (config->swaybg_client == NULL) {
+		sway_log_errno(SWAY_ERROR, "wl_client_create failed");
+		return false;
+	}
+
+	config->swaybg_client_destroy.notify = handle_swaybg_client_destroy;
+	wl_client_add_destroy_listener(config->swaybg_client,
+		&config->swaybg_client_destroy);
+
+	pid_t pid = fork();
+	if (pid < 0) {
+		sway_log_errno(SWAY_ERROR, "fork failed");
+		return false;
+	} else if (pid == 0) {
+		pid = fork();
+		if (pid < 0) {
+			sway_log_errno(SWAY_ERROR, "fork failed");
+			_exit(EXIT_FAILURE);
+		} else if (pid == 0) {
+			if (!set_cloexec(sockets[1], false)) {
+				_exit(EXIT_FAILURE);
+			}
+
+			char wayland_socket_str[16];
+			snprintf(wayland_socket_str, sizeof(wayland_socket_str),
+				"%d", sockets[1]);
+			setenv("WAYLAND_SOCKET", wayland_socket_str, true);
+
+			execvp(command[0], command);
+			sway_log_errno(SWAY_ERROR, "execvp failed");
+			_exit(EXIT_FAILURE);
+		}
+		_exit(EXIT_SUCCESS);
+	}
+
+	if (close(sockets[1]) != 0) {
+		sway_log_errno(SWAY_ERROR, "close failed");
+		return false;
+	}
+	if (waitpid(pid, NULL, 0) < 0) {
+		sway_log_errno(SWAY_ERROR, "waitpid failed");
+		return false;
+	}
+
+	return true;
+}
+
+bool spawn_swaybg(void) {
+	if (!config->swaybg_command) {
+		return true;
+	}
+
+	size_t length = 2;
+	for (int i = 0; i < config->output_configs->length; i++) {
+		struct output_config *oc = config->output_configs->items[i];
+		if (!oc->background) {
+			continue;
+		}
+		if (strcmp(oc->background_option, "solid_color") == 0) {
+			length += 4;
+		} else if (oc->background_fallback) {
+			length += 8;
+		} else {
+			length += 6;
+		}
+	}
+
+	char **cmd = calloc(1, sizeof(char **) * length);
+	if (!cmd) {
+		sway_log(SWAY_ERROR, "Failed to allocate spawn_swaybg command");
+		return false;
+	}
+
+	size_t i = 0;
+	cmd[i++] = config->swaybg_command;
+	for (int j = 0; j < config->output_configs->length; j++) {
+		struct output_config *oc = config->output_configs->items[j];
+		if (!oc->background) {
+			continue;
+		}
+		if (strcmp(oc->background_option, "solid_color") == 0) {
+			cmd[i++] = "-o";
+			cmd[i++] = oc->name;
+			cmd[i++] = "-c";
+			cmd[i++] = oc->background;
+		} else {
+			cmd[i++] = "-o";
+			cmd[i++] = oc->name;
+			cmd[i++] = "-i";
+			cmd[i++] = oc->background;
+			cmd[i++] = "-m";
+			cmd[i++] = oc->background_option;
+			if (oc->background_fallback) {
+				cmd[i++] = "-c";
+				cmd[i++] = oc->background_fallback;
+			}
+		}
+	}
+
```

Comment:
Can we assert that we've not assigned past the array end here?

Reply:
Done